### PR TITLE
#96 more data for trigger payload

### PR DESCRIPTION
**Pull request by CarlosNZ**
This is not on the immediate priority list, but I think it will make writing front-end queries for plugins substantially easier going forward, and it kind of goes hand-in-hand with the plugin output format from last week.

Have also substantially updated the documentation page (Triggers & Actions) to explain in more detail about how information is passed to Actions for evaluation, etc. So I'd recommending reading that new documentation first.

**Also**: made a fix to the stage/status computed columns functions, as it wasn't taking into account the `is_current` value.

### CODE REVIEW

**Review by andreievg — CHANGES_REQUESTED** [review]
A few comment, mainly regarding adding logic in DB layer.

Since improvements are added in this PR, I want to my thoughts about payloads (correct me if I am wrong, and please share you reflection on below)

We have trigger needing some overall data for evaluating if actions needs to be run.
We have actions, that require some parameters.

Both triggers and actions use evaluator, and require 'core' to provide some information to evaluate against. 
I like the idea of generic payload for to be available in evaluator (with the same shape), for both trigger and actions.

I think actions should get the payload of 'current' state in the payload (not the historic state at the time things were triggered), i.e. do it in executeAction.
So as per above, we don't technically need to application_data and parameters_evaluated in the action table.

We want to simplify evaluator expressions, and I think that this generic payload, should have 90% of all possible things that would be needed for parameters and trigger conditions.

Thus we can have a method that can compile this generic payload (we can add things to the payload as we develop core plugins), basically what you've done with `getTriggerPayloadData` (except as per comment further down, not tightly coupled with DB class) 

I think action 'outcomes' can become very hard to keep track of (i am still quite resiliant to rellying on outputs on previous sequenced actions).

So for the example of username (in cLog, userRegistration), all of the responses should be available to the action, and the query should be something like:

```
 operator: "CONCAT"
    children: [
        { value: "Output concatenation: The user " }
        {
            operator: "objectProperties"
            children: [
                { value: { property: "applicationResponses.username.value" } }
                ]
        }
        { value: "'s registration has been " }
        {
            operator: "objectProperties"
            children: [
                { value: { property: "applicationInfo.outcome" } }
            ]
        }
    ]
```

File: src/components/postgresConnect.ts [hunk 1]
```diff
@@ -218,7 +219,54 @@ class PostgresDB {
...
+      stage_id, stage_number, stage,
+      stage_history_id, stage_history_time_created,
+      status_history_id, status, status_history_time_created
+      FROM application_stage_status_all
+      WHERE application_id = $1
+      AND stage_is_current = true
+      AND status_is_current = true
+    `
+    const result = await this.query({ text, values: [applicationId] })
+    if (result.rows.length > 1) throw new Error('Database inconsistency')
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/postgresConnect.ts:235]: So just remove the line you think?

File: src/components/postgresConnect.ts [hunk 2]
```diff
@@ -218,7 +219,54 @@ class PostgresDB {
...
+    const dbOutput = result.rows[0]
+    const applicationData = {
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/postgresConnect.ts:237]: Cos if you name fields with camelCase in SELECT, SQL will return them all lower-case.
So `SELECT stage_id as stageId` just returns `{ stageid: xx } ` :(

**Comment by CarlosNZ** [comment]
Okay, have done as suggested and discussed, and [made a new card](https://github.com/orgs/openmsupply/projects/6#card-49189086) for remaining (as discussed)

- new module `triggerFetchData` to fetch and build application data (where available), including all responses and user info.
- Removed `createUserFromApplication` plugin, since it's now easy to just use the `createUser` plugin with simpler queries.
- Updated `insertData.js` with slightly new Action definitions

File: src/components/postgresConnect.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/components/postgresConnect.ts:237]: yeap

**Review by andreievg — COMMENTED** [review]
Looks good thanks Carl, before approval though, did you by accident delete some actions ?

**Comment by andreievg** [comment]
Oops should have read your comment

**Review by andreievg — APPROVED** [review]

**Pull request merged**